File: leetcode-js/49.js
function groupAnagrams(strs) {
    const alphabet = "abcdefghijklmnopqrstuvwxyz";

    let groups = new Map();

    strs.forEach((str) => {
        let mask = Array.from({ length: 26 }).fill(0);

        for (let char of str) {
            mask[char.charCodeAt() - 97] += 1;
            console.log(mask[char.charCodeAt() - 97]);
        }

        let maskStr = mask.join("-");
        console.log(maskStr, mask);

        if (groups.has(maskStr)) {
            let anagrams = groups.get(maskStr);
            groups.set(maskStr, [...anagrams, str]);
        } else {
            groups.set(maskStr, [str]);
        }
    });

    let result = [];

    for (let [key, value] of groups) {
        result.push(value);
    }

    return result;
}

function groupAnagrams2(strs) {
    let groups = new Map();

    strs.forEach((str) => {
        let sortedStr = Array.from(str).sort().join("");

        if (groups.has(sortedStr)) {
            let anagrams = groups.get(sortedStr);
            groups.set(sortedStr, [...anagrams, str]);
        } else {
            groups.set(sortedStr, [str]);
        }
    });

    let result = [];

    for (let [key, value] of groups) {
        result.push(value);
    }

    return result;
}

let strs = ["bdddddddddd", "bbbbbbbbbbc"];

console.log(groupAnagrams(strs));

File: leetcode-go/49.go
package main

import (
	"sort"
)

type sortRunes []rune

func (s sortRunes) Less(i, j int) bool {
	return s[i] < s[j]
}

func (s sortRunes) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

func (s sortRunes) Len() int {
	return len(s)
}

func SortString(s string) string {
	r := []rune(s)
	sort.Sort(sortRunes(r))
	return string(r)
}

func groupAnagrams(strs []string) [][]string {
	groups := make(map[string][]string)

	for _, str := range strs {
		sortedStr := SortString(str)
		groups[sortedStr] = append(groups[sortedStr], str)
	}

	result := [][]string{}
	for _, group := range groups {
		result = append(result, group)
	}
	return result
}
